veps/sig-compute/hypervisor-abstraction.md

```diff
@@ -0,0 +1,590 @@
+   ```go
+    type DefaultsProvider interface {
+      ApplyVMDefaults(vm *v1.VirtualMachine, cc *virtconfig.ClusterConfig, client kubecli.KubevirtClient)
+      ApplyVMISpecDefaults(spec *v1.VirtualMachineInstanceSpec, cc *virtconfig.ClusterConfig) error
+      FinalizeVMI(vmi *v1.VirtualMachineInstance, cc *virtconfig.ClusterConfig) error
+    }
+   ```
+   Only zero-value fields are set at each layer; `FinalizeVMI` handles derived/status data (CPU topology snapshot, memory status, hotplug sizing, feature dependency resolution). Existing public functions delegate to the resolved provider for backwards compatibility.
+- **Runtime interface (`pkg/hypervisor/runtime/`)** – Provides a shared `HypervisorRuntime` contract for runtime-specific behavior such as `AdjustResources`, `HandleHousekeeping`, and `GetMemoryOverhead`.
+
+  ```go
+  type HypervisorRuntime interface {
+    AdjustResources(vmi *v1.VirtualMachineInstance, additionalOverheadRatio *string) error
+    HandleHousekeeping(vmi *v1.VirtualMachineInstance, domain *api.Domain) error
+    GetMemoryOverhead(vmi *v1.VirtualMachineInstance, arch string, additionalOverheadRatio *string) resource.Quantity
+  }
+  ```
+
+- **Converter library (`pkg/virt-launcher/virtwrap/converter/hypervisor/`)** – Implements the new `HypervisorConverter` interface described below. Each hypervisor file focuses on XML/domain differences while `base.go` holds the shared helpers. The converter selects the correct implementation via a local registry keyed by hypervisor name.
+
+  ```golang
+  // pkg/virt-launcher/virtwrap/converter/hypervisor/converter.go
+  type HypervisorConverter interface {
+      SetDomainType(domain *api.Domain, ctx *ConverterContext) error
+      ConvertWatchdog(source *v1.Watchdog, watchdog *api.Watchdog) error
+      ValidateDiskBus(bus v1.DiskBus) error
+      LaunchSecurity(vmi *v1.VirtualMachineInstance) *api.LaunchSecurity
+      SetIOThreads(vmi *v1.VirtualMachineInstance, domain *api.Domain, vcpus uint) error
+      ConvertClock(source *v1.Clock, clock *api.Clock) error
+      ConvertFeatures(source *v1.Features, features *api.Features, ctx *ConverterContext) error
+  }
+  ```
+
+- **Node labeller (`pkg/virt-handler/node-labeller/hypervisor/`)** – Adds a lightweight hook so each hypervisor can declare the devices to probe, the preferred libvirt `virt-type`, and optional feature discovery (such as Hyper-V enlightenments for KVM on amd64).
+
+This split preserves the “implement once, reuse everywhere” story without routing everything through a monolithic interface. New hypervisors can land incrementally—start with defaults and webhooks, add converter support, then extend node labelling—while keeping the contract for each area explicit and testable.
+
```

Comment:
I don't know where else to put it, as I'm not _that_ closely familiar with kubevirt or libvirt internals -- but does "custom XML for libvirt domain spec" make sense anywhere in this proposal? Supporting that might allow more immediate support for some features of custom hypervisors/virtualizers which have no direct support in Kubevirt. (Fictitious driver features: "We support iSCSI with additional tweaks to retry logic if you just throw in this custom XML" or "We support webhooks on events such as user ejecting the virtual floppy")

Reply:
@harshitgupta1337 is correct, in principle this is possible with hook sidecars today (there are some gotchas though). It should be kept out of this proposal.

Reply:
@ivucica , can this thread be resolved?